sklearn/_loss/_loss.pxd

```diff
@@ -0,0 +1,75 @@
+# cython: language_level=3
+
+import numpy as np
+cimport numpy as np
+
+np.import_array()
+
+
+# Fused types for y_true, y_pred, raw_prediction
+ctypedef fused Y_DTYPE_C:
+    np.npy_float64
+    np.npy_float32
+
+
+# Fused types for gradient and hessian
+ctypedef fused G_DTYPE_C:
+    np.npy_float64
+    np.npy_float32
+
+
+# Struct to return 2 doubles
+ctypedef struct double2:
+   double val1
+   double val2
+
+
+# C base class for loss functions
+cdef class cLossFunction:
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cHalfSquaredError(cLossFunction):
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cAbsoluteError(cLossFunction):
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cPinballLoss(cLossFunction):
+    cdef readonly double quantile  # readonly makes it inherited by children
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cHalfPoissonLoss(cLossFunction):
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cHalfGammaLoss(cLossFunction):
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cHalfTweedieLoss(cLossFunction):
+    cdef readonly double power  # readonly makes it inherited by children
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
+
+
+cdef class cBinaryCrossEntropy(cLossFunction):
+    cdef double closs(self, double y_true, double raw_prediction) nogil
+    cdef double cgradient(self, double y_true, double raw_prediction) nogil
+    cdef double2 cgrad_hess(self, double y_true, double raw_prediction) nogil
```

Comment:
Good question! `cCategoricalCrossEntropy` does not have an implementation of `closs`, `cgradient` and `cgrad_hess`. Those C-functions working on single data points are intended for SGD in linear models, and there is no multiclass support, so far.
On top, the signature would be different. We have
```Cython
cdef double closs(self, double y_true, double raw_prediction)
```
But for `cCategoricalCrossEntropy`, it would be
```Cython
cdef double closs(self, double y_true, double* raw_prediction)
```